content/courses/Creating_C_Bindings_For_Ada/chapters/08_Special_Cases.rst

```diff
@@ -2,16 +2,187 @@
+
+.. code-block:: c
+
+  struct OpenPipeT message = { 1, 2 };
+  setPipeCommand ( OPEN, &message );
+
+The concern in this case (regardless of language), is there is
+no verification that the contents at the address is a valid
+message, or even is a legal address.
+
+.. code-block:: Ada
+  :caption: Ada Thin Binding
+
+  with Interfaces.C;
+  with System;
+  package Message_H is
+     type Open_Pipe_T is record
+        Pipe_Id   : Interfaces.C.unsigned;
+        Requestor : Interfaces.C.unsigned;
+     end record with Convention => C_Pass_By_Copy;
+     type Change_Flow_T is record
+        Pipe_Id    : Interfaces.C.unsigned;
+        Requestor  : Interfaces.C.unsigned;
+        Percentage : Interfaces.C.short;
+     end record with Convention => C_Pass_By_Copy;
+     type Close_Pipe_T is record
+        Pipe_Id   : Interfaces.C.unsigned;
+        Requestor : Interfaces.C.unsigned;
+     end record with Convention => C_Pass_By_Copy;
+     type Command_T is (Open, Change, Close) with Convention => C;
+     procedure Set_Pipe_Command (Command : Command_T;
+                                 Content : System.Address)
+       with Import => Standard.True,
+            Convention => C,
+            External_Name => "setPipeCommand";
+  end Message_H;
+
+The thin binding of the header file will allow us to do the same -
+create an object, and pass the address of the object into
+:ada:`Set_Pipe_Command`.
+
+.. code-block:: Ada
+
+   procedure Open is
+      Message : Open_Pipe_T := (1, 2);
+   begin
+      Set_Pipe_Command (Open, Message'address);
+   end Open;
+
+But this still leaves the concern about random addresses.
+
+If we were writing this in Ada to begin with, one solution would be to
+use generics (either a subprogram or package) to implement different versions
+of :ada:`Set_Pipe_Command` that would take a parameter of the appropriate
+message type and have it call the thin binding, remove the ability of the
+clients to abuse the address parameter.
+
+.. code-block:: Ada
+  :caption: Ada Thick Binding
+
+   generic
+      Command : Command_T;
+      type Content_T is private;
+   procedure Generic_Set_Pipe_Command (Content : Content_T);
+
+   procedure Generic_Set_Pipe_Command (Content : Content_T) is
+   begin
+      Set_Pipe_Command (Command, Content'address);
+   end Generic_Set_Pipe_Command;
+
+.. code-block:: Ada
+  :caption: Ada Thick Binding Usage
+
+   procedure Open_Command is new Generic_Set_Pipe_Command
+     (Open, Open_Pipe_T);
+   procedure Change_Flow_Command is new Generic_Set_Pipe_Command
+     (Change, Change_Flow_T);
+   procedure Close_Command is new Generic_Set_Pipe_Command
+     (Close, Close_Pipe_T);
+
+   procedure Example is
+      Open_Pipe : Open_Pipe_T := (1, 2);
+      Change_Flow : Change_Flow_T := (1, 2, 3);
+      Close_Pipe : Close_Pipe_T := (1, 2);
+   begin
+      Open_Command (Open_Pipe);
+      Change_Flow_Command (Change_Flow);
+      Close_Command (Close_Pipe);
+   end Example;
```

Comment:
Do we leave `procedure Set_Pipe_Command` accessible in the thick binding version?

Reply:
For people not used to strong typing, what makes the instantiation of the generic get rid of the "random addresses" problem? I would explain the distinction more explicitly.